samples/common/xbee3_ble_scanner.c

```diff
@@ -0,0 +1,293 @@
+
+#include <stdio.h>
+#include <limits.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "xbee/platform.h"
+#include "xbee/device.h"
+#include "xbee/ble.h"
+
+#include "_xbee_term.h"  
+
+
+typedef enum ble_scan_state_e {
+    SCAN_STATE_INIT,
+    SCAN_STATE_AWAIT_START,
+    SCAN_STATE_RUNNING,
+    SCAN_STATE_STOPPED,
+    SCAN_STATE_ERROR,
+} ble_scan_state_t;
+
+static ble_scan_state_t scan_state;
+
+#define XBEE_BLE_FRAME_HANDLERS \
+    { XBEE_FRAME_BLE_SCAN_STATUS, 0, scan_status_handler, NULL}, \
+    { XBEE_FRAME_BLE_SCAN_LEGACY_ADVERT_RESP, 0, xbee_ble_scan_legacy_advert_dump, NULL}, \
+    { XBEE_FRAME_BLE_SCAN_EXT_ADVERT_RESP, 0, xbee_ble_scan_ext_advert_dump, NULL}
+
+#define BLE_SCAN_INTERVAL_DEFAULT_USEC (0x138800) // 1.28 seconds
+#define BLE_SCAN_WINDOW_DEFAULT_USEC   (0x2BF2)   // 11.25 milliseconds
+
+static int scan_status_handler(xbee_dev_t *xbee, const void FAR *raw,
+                               uint16_t length, void FAR *context)
+{
+    const xbee_frame_ble_scan_status_t FAR *status = raw;
+
+    switch (status->scan_status) {
+    case XBEE_BLE_SCAN_STATUS_SUCCESS:
+        /* Intentional fall-through */
+    case XBEE_BLE_SCAN_STATUS_RUNNING:
+        scan_state = SCAN_STATE_RUNNING;
+        fprintf(stderr, "Scan is running\n\n");
+        break;
+
+    case XBEE_BLE_SCAN_STATUS_STOPPED:
+        scan_state = SCAN_STATE_STOPPED;
+        fprintf(stderr, "Scan has stopped\n");
+        break;
+
+    case XBEE_BLE_SCAN_STATUS_ERROR:
+        fprintf(stderr, "The scan failed to start\n");
+        scan_state = SCAN_STATE_ERROR;
+        break;
+    case XBEE_BLE_SCAN_STATUS_INVALID_PARAMS:
+        fprintf(stderr, "The scan failed to start due to invalid parameters\n");
+        scan_state = SCAN_STATE_ERROR;
+        break;
+    default:
+        fprintf(stderr, "An unknown error occoured: %" PRIu8"\n",
+                status->scan_status);
+        scan_state = SCAN_STATE_ERROR;
+        break;
+   };
+
+    return 0;
+}
+
+int ble_scanner_sample(xbee_dev_t *xbee, uint16_t scan_duration,
+                       uint32_t scan_window, uint32_t scan_interval,
+                       const char *filter_local_name)
+{
+    int result;
+    const uint8_t filter_type = filter_local_name ? XBEE_BLE_SCAN_FILTER_NAME : 0;
+    const size_t filter_len = filter_local_name ? strlen(filter_local_name) : 0;
+    scan_state = SCAN_STATE_INIT;
+
+    while (scan_state < SCAN_STATE_STOPPED) {
+        result = xbee_dev_tick(xbee);
+        if (result < 0) {
+            fprintf(stderr, "xbee_dev_tick failed with %d\n", result);
+            return EXIT_FAILURE;
+        }
+
+        if (scan_state == SCAN_STATE_INIT) {
+            // Send scan initiation
+            fprintf(stderr, "Initiating scan\n");
+            result = xbee_ble_scan_start(xbee, scan_duration, scan_window,
+                        scan_interval, filter_type,
+                        (const uint8_t *)filter_local_name, filter_len);
```

Comment:
As mentioned in a previous comment, updating the `xbee_ble_scan_start()` API to use `const void *` would eliminate the need for this cast.